### ISSUE TITLE
Recognize faces realtime video

### ISSUE THREAD

**Issue by Zumbalamambo** [davidsandberg/facenet#437]
How do I recognize faces realtime?

**Comment by a73013110** [comment]
I think it's not possible to recognize in real time, because the pre-train model is large.
After the test, embedding image in win10 may take 3 second with official installation, seem that cmake tensorflow with SIMD and GPU will get 5 times fast but I didn't cmake successfully.
And I hear that in Linux may take 0.5 second.
By the way, I use GTX 1060 and tensorflow-gpu for test

**Comment by mingfengwuye** [comment]
Based on this, i get video from a RTSP camera, and send every one of three images to facenet and use knn to recognize. After some optimization, it is almost real-time. I use Titan X.

**Comment by pavankumarkatakam** [comment]
Hi @mingfengwuye , can you please tell me how many frames per second you are processing with titanx with settings like minsize 20, image size 160, margin 44 .. I only have nvidia 960m working on hd video (1280*720) with above settings i got actually 2 frames per second (provided all screens has faces). for human 2fps is watching like slow motion.

**Comment by vudung45** [comment]
@pavankumarkatakam @Zumbalamambo, so I composed everything into a simple repo
https://github.com/vudung45/FaceRec
This is a simple minified source of a bigger project I was working on this summer. Basically get images from VideoCapture -> detect faces -> label them in real time.
Check it out if your interested! Fork or star if you like it :)

**Comment by pavankumarkatakam** [comment]
Hi @vudung45 , I have seen the repo and tested. Some points I noted are
1. Your are using the DLIB like alignment, where as MTCNN is only for face detection.
2.  While detecting faces you have taken a scale factor how it is different from factor=0.709?
3. Lastly you are generating the embedding like below
`self.embeddings = tf.nn.l2_normalize(resnet.inference(self.x, 0.6, phase_train=False)[0], 1, 1e-10);` but in this repo embeddings are generated like `embeddings = tf.get_default_graph().get_tensor_by_name("embeddings:0")` what is the difference between them?

**Comment by vudung45** [comment]
@pavankumarkatakam 
1. I indeed implemented the DLIB like strategy to align but it's not exactly the same. I find this method is somewhat more reliable than the Dlib's one since it doesn't deform the original image.

2. The 0.709 factor is the image pyramid magic number variable, you don't have to care about it.  scale_factor, on the other hand, is a resizing variable that I declare to determine how much you want to resize the original frame before trying to find faces. This will help speeding up the detection speed, however, it will decrease chances of detecting small faces ( 2 - 4 is the optimal number here, depending on how big is your original frame).

3. Since I'm not using the frozen graph, my source extracts the face feature from the architecture source code. You can take a look at the resnet_model source code to know what I'm talking about. Other than that, there is no difference between those 2 approaches.

**Comment by pavankumarkatakam** [comment]
Thanks @vudung45 for explanation.

**Comment by pavankumarkatakam** [comment]
Hi @vudung45 , Does if we resize the image using the scale factor in MTCNNDetect some faces will miss right?

**Comment by pavankumarkatakam** [comment]
@vudung45, Did you check with and without aligning using "AlignCustom" , Is there any improvement when aligning the faces.

**Comment by vudung45** [comment]
@pavankumarkatakam there is a decent improvement when you align the inputs. The whole goal of aligning is to make your input as "simple/easy" as possible.
And yes, if you increase scale_factor, it will be more likely to miss the faces on screen. That's why I mentioned you have to find an optimal scale_factor for your image resolution.

**Comment by davidsandberg** [comment]
Closing due to inactivity. Reopen if needed.

**Comment by SehgalDivij** [comment]
@davidsandberg Is it possible to anonymously recognize faces on the go as they enter a video stream? What I mean is say I want to catalogue all faces into my system the first time they appear and the next time they appear in a video stream, I want to recognize that a person had already appeared before.

Is it possible to utilize the embeddings by doing some form of clustering?

Another totally unrelated question: Is it possible to estimate the angle from which the face was seen, when an embedding is generated from a face?

**Comment by deepconsc** [comment]
@SehgalDivij for the first question - kinda yes. By extracting embeddings for every detected face and appending them in case they're not presented in array (i.e. distance between new embedding and the old ones is above the threshold), but that'll take huge computational resources.
